### PR TITLE
feature/search-modal

### PR DESCRIPTION
**Pull request by Kacper-RF**
- [x] Component with container
- [x] Base story
- [x] MUI
- [x] Transparent modal  with logic
- [x] Base search logic(list all available channels)
- [x] Key combo to open (ctrl +k and ctrl+t)
- [x] Close on escape
- [x] Close on icon click
- [x] Change channel on click and close modal
- [x] Input styles
- [x] Highlight result (channel)
- [x] Show recent channels
- [x] Component test
- [x] Change channel on press enter 
- [x] Navigate by arrows on channel list
- [x] Component for single channel item
- [x] RTL test
- [x] Focus Bug
- [x] Create selector with all logic to filter channels
- [x] Unread channel logic in component
- [x] Set focus on first channel if is unread in other case set focus on input
- [x] RTL test for unread channel
- [x] Change text if channel is unread
- [x] Fix bug with key in map

image

Unread message feature:

image

### CODE REVIEW

File: packages/desktop/src/renderer/components/SearchModal/SearchModal.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import { useModal } from '../../containers/hooks'
+import { ModalName } from '../../sagas/modals/modals.types'
+
+const SearchModal = () => {
+  const dispatch = useDispatch()
+  const searchChannelModal = useModal(ModalName.searchChannelModal)
+
+  const publicChannelsSelector = useSelector(publicChannels.selectors.publicChannels)
+
+  const setCurrentChannel = (address: string) => {
```
Comment (siepra) [inline packages/desktop/src/renderer/components/SearchModal/SearchModal.tsx:15]: useCallback?

**Review by siepra — APPROVED** [review]

File: packages/desktop/src/renderer/components/SearchModal/ChannelItem.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+const ChannelItem = ({
+  item,
+  focused,
+  className,
+  classNameSelected,
+  onClickHandler,
+  onKeyPressHandler
+}) => {
+  const [setted, isSetted] = useState(false)
```
Comment (siepra) [inline packages/desktop/src/renderer/components/SearchModal/ChannelItem.tsx:13]: It's kind of silly but there's no such word as 'setted'

File: packages/desktop/src/renderer/components/SearchModal/SearchModelComponent.tsx [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+  }
+
+  const closeHandler = () => {
+    setCurrentFocus(null)
+    setChannelInput('')
+    handleClose()
+  }
+
+  const ref = useRef<HTMLDivElement>()
+  // useEffect(() => {
```
Comment (siepra) [inline packages/desktop/src/renderer/components/SearchModal/SearchModelComponent.tsx:184]: leftover?

File: packages/desktop/src/renderer/components/SearchModal/SearchModelComponent.tsx [hunk 4]
```diff
@@ -0,0 +1,280 @@
...
+  const [focusedIndex, setCurrentFocus] = useCyclingFocus(channelList.length)
+
+  const onChange = (value: string) => {
+    setChannelInput(value)
+  }
+
+  const onKeyPressHandler = (e: React.KeyboardEvent<HTMLDivElement>, address: string) => {
+    e.preventDefault()
+    if (e.key === 'Enter') {
+      setCurrentChannel(address)
```
Comment (siepra) [inline packages/desktop/src/renderer/components/SearchModal/SearchModelComponent.tsx:165]: Guess you can just call `onChannelClickHandler` here

File: packages/desktop/src/renderer/containers/hooks.ts [hunk 5]
```diff
@@ -68,9 +67,30 @@ export const useCyclingFocus = (
           }
           setCurrentFocus(focusValue)
           break
+        case 'ArrowDown':
```
Comment (siepra) [inline packages/desktop/src/renderer/containers/hooks.ts:70]: Double check if it doesn't interrupt the channel page behavior. I mean there's a risk one may have problems changing focus inside message input using keyboard's arrows

File: packages/desktop/src/renderer/containers/hooks.ts [hunk 6]
```diff
@@ -68,9 +67,30 @@ export const useCyclingFocus = (
...
+          evt.preventDefault()
+          if (currentFocus === null) {
+            focusValue = 0
+          } else if (currentFocus !== listSize - 1) {
+            focusValue = currentFocus + 1
+          }
+          setCurrentFocus(focusValue)
+          break
+        case 'ArrowUp':
+          evt.preventDefault()
```
Comment (siepra) [inline packages/desktop/src/renderer/containers/hooks.ts:80]: I think you unnecessarily duplicated the code (identical lines appears under 'PageUp' case). Maybe it's worth it to extract it into a separate function and call it in those two places

File: packages/desktop/src/rtl-tests/searchModal.test.tsx [hunk 7]
```diff
@@ -0,0 +1,183 @@
...
+    fireEvent.click(tab)
+
+    await act(async () => {})
+
+    const currentChannel = publicChannels.selectors.currentChannel(redux.store.getState())
+
+    expect(currentChannel.name).toEqual(CHANNEL_NAME)
+  })
+
+  it.skip('Select most recent channel by clicking arrow down and enter', async () => {
```
Comment (siepra) [inline packages/desktop/src/rtl-tests/searchModal.test.tsx:148]: Is it skipped in purpose?

File: packages/state-manager/src/sagas/publicChannels/publicChannels.selectors.ts [hunk 8]
```diff
@@ -73,6 +73,28 @@ export const currentChannelAddress = createSelector(
   }
 )
 
+export const dynamicSearchedChannels = (channelInput: string) =>
+  createSelector(publicChannels, publicChannelsSelector => {
+    const recentChannels = publicChannelsSelector
```
Comment (siepra) [inline packages/state-manager/src/sagas/publicChannels/publicChannels.selectors.ts:78]: I wish it was a separate selector. Try to make them as atomic as possible. See eg. `displayableCurrentChannelMessages` - it reuses `sortedCurrentChannelMessages` which reuses `currentChannelMessages` etc... thanks to that we get the best of memoization in selectors.

File: packages/state-manager/src/sagas/publicChannels/publicChannels.selectors.ts [hunk 9]
```diff
@@ -73,6 +73,28 @@ export const currentChannelAddress = createSelector(
...
+      .sort((a, b) => b.timestamp - a.timestamp)
+      .slice(0, 3)
+
+    const generalChannel = publicChannelsSelector.find(channel => channel.name === 'general')
```
Comment (siepra) [inline packages/state-manager/src/sagas/publicChannels/publicChannels.selectors.ts:82]: I think you can have a separate selector for general channel as well

File: packages/state-manager/src/sagas/publicChannels/publicChannels.selectors.ts [hunk 10]
```diff
@@ -73,6 +73,28 @@ export const currentChannelAddress = createSelector(
...
+
+    const filteredList = publicChannelsSelector.filter(channel =>
+      channel.name.includes(channelInput)
+    )
+
+    const isFilteredList = filteredList.length > 0 ? filteredList : recentChannels
+
+    const areThreeRecentChannels = recentChannels.length >= 3 ? recentChannels : [generalChannel]
```
Comment (siepra) [inline packages/state-manager/src/sagas/publicChannels/publicChannels.selectors.ts:90]: You can include this condition in the selector I mentioned before. recentChannels selector could return either a list of channels or simply a general channel

**Comment by pdurbin** [comment]
It works! Thank you! 🎉 

Screen Shot 2023-01-17 at 7 13 41 PM

I'm not sure of your process but perhaps you could have this issue close automatically when this PR is merged:

- #544

File: packages/desktop/src/renderer/containers/hooks.ts (hunk 5 above)
Comment (Kacper-RF) [reply packages/desktop/src/renderer/containers/hooks.ts]: Yes there was a bug, I wrote some workaround to handle the logic correctly, please take a look at this new part of the code.

File: packages/desktop/src/rtl-tests/searchModal.test.tsx [hunk 11]
```diff
@@ -181,4 +186,60 @@ describe('Switch channels', () => {
     await userEvent.type(text, '{escape}')
     expect(text).not.toBeVisible()
   })
+
+  it('Unread message', async () => {
```
Comment (siepra) [inline packages/desktop/src/rtl-tests/searchModal.test.tsx:190]: I wish it was more descriptive

**Review by siepra — APPROVED** [review]

**Pull request merged**